SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/Visitor.h

```diff
@@ -235,8 +230,79 @@ class SOFA_SIMULATION_CORE_API Visitor
     static void dumpInfo( const std::string &info);
 #endif
 };
-} // namespace simulation
 
-} // namespace sofa
+template< class VisitorType, class VContext, class ObjectType>
+void Visitor::runVisitorTask(VisitorType *visitor,
+                             VContext *ctx,
+                             void (VisitorType::*task)(VContext *, ObjectType *),
+                             ObjectType *ptr,
+                             const std::string &typeInfo)
+{
+    if(this->testTags(ptr))
+    {
+#ifdef SOFA_VERBOSE_TRAVERSAL
+        visitor->debug_write_state_before(ptr);
+#endif
+        auto t = this->begin(ctx, ptr, typeInfo);
+        (visitor->*task)(ctx, ptr);
+        this->end(ctx, ptr, t);
+#ifdef SOFA_VERBOSE_TRAVERSAL
+        visitor->debug_write_state_after(ptr);
+#endif
+    }
+}
+
+template< class VisitorType, class VContext, class ObjectType>
+Visitor::Result Visitor::runVisitorTask(VisitorType *visitor,
+                                        VContext *ctx,
+                                        Result (VisitorType::*task)(VContext *, ObjectType *),
+                                        ObjectType *ptr,
+                                        const std::string &typeInfo)
+{
+    Result res = Result::RESULT_CONTINUE;
+    if(this->testTags(ptr))
+    {
+#ifdef SOFA_VERBOSE_TRAVERSAL
+        visitor->debug_write_state_before(ptr);
+#endif
+        auto t = this->begin(ctx, ptr, typeInfo);
+        res = (visitor->*task)(ctx, ptr);
+        this->end(ctx, ptr, t);
+#ifdef SOFA_VERBOSE_TRAVERSAL
+        visitor->debug_write_state_after(ptr);
+#endif
+    }
+    return res;
+}
+
+template < class VisitorType, class VContext, class Container, typename PointedType >
+void Visitor::for_each(VisitorType *visitor,
+                       VContext *ctx,
+                       const Container &list,
+                       void (VisitorType::*task)(VContext *, PointedType *),
+                       const std::string &typeInfo)
+{
+    for (const auto& element : list)
+    {
+        runVisitorTask(visitor, ctx, task, &*element, typeInfo);
+    }
+}
+
+template < class VisitorType, class VContext, class Container, typename PointedType >
+Visitor::Result Visitor::for_each(VisitorType *visitor,
+                                  VContext *ctx,
+                                  const Container &list,
+                                  Visitor::Result (VisitorType::*task)(VContext *, PointedType *),
+                                  const std::string &typeInfo)
+{
+    Visitor::Result res = Visitor::RESULT_CONTINUE;
+    for (const auto& element : list)
+    {
+        res = runVisitorTask(visitor, ctx, task, &*element, typeInfo);
+    }
+    return res;
+}
+
+} // namespace sofa::simulation
 
 #endif
```

Comment:
if you go for a clean, you may remove the ifdef/endif guard with a pragma

Reply:
done